{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "87e3b62e-1faf-4f2b-aed7-76f33f18560f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# import relevant packages\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import scipy.stats as ss\n",
    "import seaborn as sns\n",
    "\n",
    "import torch\n",
    "import torch_geometric\n",
    "from torch_geometric.loader import DataLoader\n",
    "\n",
    "# set figure parameters\n",
    "sc.settings.set_figure_params(dpi=100)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1aa844fd-a75c-4442-ae59-38f17cf1c8db",
   "metadata": {},
   "source": [
    "## define helper functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "56464c53-2e89-4bd5-b19f-28ffe6dea02c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# define a function to gather positions\n",
    "def get_pos(n_x, n_y):\n",
    "    # create the hex lattice\n",
    "    xs = np.array([np.arange(0, n_x) + 0.5 if idx % 2 == 0 else np.arange(0, n_x) for idx in range(n_y)])\n",
    "    # derive the y-step given a distance of one\n",
    "    y_step = np.sqrt(1**2+0.5**2)\n",
    "    ys = np.array([[y_step * idy] * n_x for idy in range(n_y)])\n",
    "    # define the positions\n",
    "    pos = np.vstack([xs.flatten(), ys.flatten()]).T\n",
    "    return pos"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7dd3e0be-624c-4129-80c2-9b52bc5cc9f8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# define a function to derive the gex from the sphex\n",
    "def calc_gex(sphex):\n",
    "    \"\"\"\n",
    "    Calculates the gene expression matrix from the spherical\n",
    "    \"\"\"\n",
    "    # setup the gex\n",
    "    n_genes = sphex.shape[1]+1\n",
    "    gex = torch.from_numpy(np.zeros((sphex.shape[0], n_genes)).astype('float32'))\n",
    "    # compute the gex\n",
    "    for idx in range(n_genes):\n",
    "        if idx == n_genes-1:\n",
    "            gex[:,idx] = torch.sin(sphex[:,idx-1])\n",
    "        else:\n",
    "            gex[:,idx] = torch.cos(sphex[:,idx])\n",
    "        for idx_ in range(idx):\n",
    "            gex[:,idx] *= torch.sin(sphex[:,idx_])\n",
    "    return gex"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "47d9fddf-d8b9-49c3-ad25-6a4202f33df5",
   "metadata": {},
   "outputs": [],
   "source": [
    "from torch_geometric.nn import GCNConv\n",
    "from sklearn.neighbors import kneighbors_graph\n",
    "# define the number of neighbors (six for visium)\n",
    "n_neighbors = 6\n",
    "# define the simcomen class\n",
    "class simcomen(torch.nn.Module):\n",
    "    # define initialization function\n",
    "    def __init__(self, input_dim, output_dim, n_neighbors, seed=0):\n",
    "        super(simcomen, self).__init__()\n",
    "        # define the seed\n",
    "        torch.manual_seed(seed)\n",
    "        # set up the graph convolution\n",
    "        self.conv1 = GCNConv(input_dim, output_dim, add_self_loops=False)\n",
    "        # set up the linear layer for intracellular gene regulation\n",
    "        self.lin = torch.nn.Linear(input_dim, output_dim)\n",
    "        # define the neighbors\n",
    "        self.n_neighbors = n_neighbors\n",
    "        # define a tracking variable for the gene expression x matrix\n",
    "        self.sphex = None\n",
    "        self.gex = None\n",
    "\n",
    "    # define a function to artificially set the g2g matrix\n",
    "    def set_g2g(self, g2g):\n",
    "        \"\"\"\n",
    "        Artifically sets the core g2g matrix to be a specified interaction matrix\n",
    "        \"\"\"\n",
    "        # set the weight as the input\n",
    "        self.conv1.lin.weight = torch.nn.Parameter(g2g, requires_grad=False)\n",
    "        # and then set the bias as all zeros\n",
    "        self.conv1.bias = torch.nn.Parameter(torch.from_numpy(np.zeros(n_genes).astype('float32')), requires_grad=False)\n",
    "\n",
    "    # define a function to artificially set the g2g matrix\n",
    "    def set_g2g_intra(self, g2g_intra):\n",
    "        \"\"\"\n",
    "        Artifically sets the core g2g intracellular matrix to be a specified matrix\n",
    "        \"\"\"\n",
    "        # set the weight as the input\n",
    "        self.lin.weight = torch.nn.Parameter(g2g_intra, requires_grad=False)\n",
    "        # and then set the bias as all zeros\n",
    "        self.lin.bias = torch.nn.Parameter(torch.from_numpy(np.zeros(len(g2g_intra)).astype('float32')), requires_grad=False)\n",
    "\n",
    "    # define a function to artificially set the sphex matrix\n",
    "    def set_sphex(self, sphex):\n",
    "        \"\"\"\n",
    "        Artifically sets the current sphex matrix\n",
    "        \"\"\"\n",
    "        self.sphex = torch.nn.Parameter(sphex, requires_grad=True)\n",
    "        \n",
    "    # define the forward pass\n",
    "    def forward(self, edge_index, batch):\n",
    "        \"\"\"\n",
    "        Forward pass for prediction or training,\n",
    "        convolutes the input by the expected interactions and returns log(Z_mft)\n",
    "        \"\"\"\n",
    "        # compute the gex\n",
    "        self.gex = calc_gex(self.sphex)\n",
    "        # compute the message\n",
    "        msg = self.conv1(self.gex, edge_index)\n",
    "        # compute intracellular message\n",
    "        msg_intra = self.lin(self.gex)\n",
    "        # compute the log z mft\n",
    "        log_z_mft = self.log_Z_mft(edge_index, batch)\n",
    "        return msg, msg_intra, log_z_mft\n",
    "\n",
    "    # define approximation function\n",
    "    def log_Z_mft(self, edge_index, batch):\n",
    "        \"\"\"\n",
    "        Mean Field Theory approximation to the partition function. Assumptions used are:\n",
    "        - expression of values of genes are close to their mean values over the visium slide\n",
    "        - \\sum_b g_{a,b} m^b >0 \\forall a, where m is the mean gene expression and g is the gene-gene\n",
    "          interaction matrix.\n",
    "        \"\"\"\n",
    "        # retrieve number of spots\n",
    "        num_spots = self.gex.shape[0]\n",
    "        # calculate mean gene expression        \n",
    "        mean_genes = torch.mean(self.gex, axis=0).reshape(-1,1)  # the mean should be per connected graph\n",
    "        # calculate the norm of the sum of mean genes\n",
    "        g = torch.norm(torch.mm( self.n_neighbors*self.conv1.lin.weight + 2*self.lin.weight, mean_genes))   # maybe needs to change to g = torch.norm(torch.mm(mean_genes, self.conv1.lin.weight))\n",
    "        # calculate the contribution for mean values        \n",
    "        z_mean = - num_spots  * torch.mm(torch.mm(torch.t(mean_genes), self.lin.weight + 0.5 * self.n_neighbors * self.conv1.lin.weight),  mean_genes)\n",
    "        # calculate the contribution gene interactions\n",
    "        z_interaction = self.z_interaction(num_spots=num_spots, g=g)\n",
    "        # add the two contributions        \n",
    "        log_z_mft = z_mean + z_interaction\n",
    "        return log_z_mft\n",
    "\n",
    "    def z_interaction(self, num_spots, g):\n",
    "        \"\"\"\n",
    "        Avoid exploding exponentials by returning an approximate interaction term for the partition function.\n",
    "        \"\"\"\n",
    "        if g>20:\n",
    "            z_interaction = num_spots * ( g - torch.log( g) )\n",
    "        else:\n",
    "            z_interaction = num_spots * torch.log((torch.exp( g) - torch.exp(- g))/( g))\n",
    "        return z_interaction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "024116b8-91ec-4d9a-b746-b9193d6e43f5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# define the celcomen class\n",
    "class celcomen(torch.nn.Module):\n",
    "    # define initialization function\n",
    "    def __init__(self, input_dim, output_dim, n_neighbors, seed=0):\n",
    "        super(celcomen, self).__init__()\n",
    "        # define the seed\n",
    "        torch.manual_seed(seed)\n",
    "        # set up the graph convolution\n",
    "        self.conv1 = GCNConv(input_dim, output_dim, add_self_loops=False)\n",
    "        # set up the linear layer for intracellular gene regulation\n",
    "        self.lin = torch.nn.Linear(input_dim, output_dim)\n",
    "        # define the neighbors\n",
    "        self.n_neighbors = n_neighbors\n",
    "        # define a tracking variable for the gene expression x matrix\n",
    "        self.gex = None\n",
    "\n",
    "    # define a function to artificially set the g2g matrix\n",
    "    def set_g2g(self, g2g):\n",
    "        \"\"\"\n",
    "        Artifically sets the core g2g matrix to be a specified interaction matrix\n",
    "        \"\"\"\n",
    "        # set the weight as the input\n",
    "        self.conv1.lin.weight = torch.nn.Parameter(g2g, requires_grad=True)\n",
    "        # and then set the bias as all zeros\n",
    "        self.conv1.bias = torch.nn.Parameter(torch.from_numpy(np.zeros(len(g2g)).astype('float32')), requires_grad=False)\n",
    "\n",
    "    # define a function to artificially set the g2g matrix\n",
    "    def set_g2g_intra(self, g2g_intra):\n",
    "        \"\"\"\n",
    "        Artifically sets the core g2g intracellular matrix to be a specified matrix\n",
    "        \"\"\"\n",
    "        # set the weight as the input\n",
    "        self.lin.weight = torch.nn.Parameter(g2g_intra, requires_grad=True)\n",
    "        # and then set the bias as all zeros\n",
    "        self.lin.bias = torch.nn.Parameter(torch.from_numpy(np.zeros(len(g2g_intra)).astype('float32')), requires_grad=False)\n",
    "\n",
    "    # define a function to artificially set the sphex matrix\n",
    "    def set_gex(self, gex):\n",
    "        \"\"\"\n",
    "        Artifically sets the current sphex matrix\n",
    "        \"\"\"\n",
    "        self.gex = torch.nn.Parameter(gex, requires_grad=False)\n",
    "        \n",
    "    # define the forward pass\n",
    "    def forward(self, edge_index, batch):\n",
    "        \"\"\"\n",
    "        Forward pass for prediction or training,\n",
    "        convolutes the input by the expected interactions and returns log(Z_mft)\n",
    "        \"\"\"\n",
    "        # compute the message\n",
    "        msg = self.conv1(self.gex, edge_index)\n",
    "        # compute intracellular message\n",
    "        msg_intra = self.lin(self.gex)\n",
    "        # compute the log z mft\n",
    "        log_z_mft = self.log_Z_mft(edge_index, batch)\n",
    "        return msg, msg_intra, log_z_mft\n",
    "\n",
    "    # define approximation function\n",
    "    def log_Z_mft(self, edge_index, batch):\n",
    "        \"\"\"\n",
    "        Mean Field Theory approximation to the partition function. Assumptions used are:\n",
    "        - expression of values of genes are close to their mean values over the visium slide\n",
    "        - \\sum_b g_{a,b} m^b >0 \\forall a, where m is the mean gene expression and g is the gene-gene\n",
    "          interaction matrix.\n",
    "        \"\"\"\n",
    "        # retrieve number of spots\n",
    "        num_spots = self.gex.shape[0]\n",
    "        # calculate mean gene expression        \n",
    "        mean_genes = torch.mean(self.gex, axis=0).reshape(-1,1)  # the mean should be per connected graph\n",
    "        # calculate the norm of the sum of mean genes\n",
    "        g = torch.norm(torch.mm( self.n_neighbors*self.conv1.lin.weight + 2*self.lin.weight, mean_genes))   # maybe needs to change to g = torch.norm(torch.mm(mean_genes, self.conv1.lin.weight))\n",
    "        # calculate the contribution for mean values        \n",
    "        z_mean = - num_spots  * torch.mm(torch.mm(torch.t(mean_genes), self.lin.weight + 0.5 * self.n_neighbors * self.conv1.lin.weight),  mean_genes)\n",
    "        # calculate the contribution gene interactions\n",
    "        z_interaction = self.z_interaction(num_spots=num_spots, g=g)\n",
    "        # add the two contributions        \n",
    "        log_z_mft = z_mean + z_interaction\n",
    "        return log_z_mft\n",
    "\n",
    "    def z_interaction(self, num_spots, g):\n",
    "        \"\"\"\n",
    "        Avoid exploding exponentials by returning an approximate interaction term for the partition function.\n",
    "        \"\"\"\n",
    "        if g>20:\n",
    "            z_interaction = num_spots * ( g - torch.log( g) )\n",
    "        else:\n",
    "            z_interaction = num_spots * torch.log((torch.exp( g) - torch.exp(- g))/( g))\n",
    "        return z_interaction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "58fe943b-3f04-4448-929a-c9c9c1288d93",
   "metadata": {},
   "outputs": [],
   "source": [
    "# define a function to normalize the g2g\n",
    "def normalize_g2g(g2g):\n",
    "    \"\"\"\n",
    "    Addresses any small fluctuations in symmetrical weights\n",
    "    \"\"\"\n",
    "    # symmetrize the values\n",
    "    g2g = (g2g + g2g.T) / 2\n",
    "    # force them to be between 0-1\n",
    "    g2g[g2g < 0] = 0\n",
    "    g2g[g2g > 1] = 1\n",
    "    # force the central line to be 1\n",
    "    for idx in range(len(g2g)):\n",
    "        g2g[idx, idx] = 1\n",
    "    return g2g"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "af41cecb-b530-4602-ad37-5b6f7b71562c",
   "metadata": {},
   "source": [
    "## read in xenium"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3d958eb7-ec9e-49d4-98e0-58841884cf05",
   "metadata": {},
   "outputs": [],
   "source": [
    "# read in xenium\n",
    "avis = sc.read_10x_h5('XENIUM_HUMAN_GBM/cell_feature_matrix.h5')\n",
    "avis.var_names = avis.var_names.astype(str)\n",
    "avis.var_names_make_unique()\n",
    "# remove cells with no counts\n",
    "sc.pp.filter_cells(avis, min_counts=100)\n",
    "# normalize the data\n",
    "sc.pp.normalize_total(avis, target_sum=1e6)\n",
    "sc.pp.log1p(avis)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e440a477-3437-459b-b02e-cb2e1a8e501f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# read in the spatial\n",
    "df = pd.read_csv('XENIUM_HUMAN_GBM/cells.csv.gz', index_col=0)\n",
    "df = df.loc[avis.obs.index]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "09f7bb08-c3c4-4612-8267-cd8425bc10c0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# view the tissue and choose a specific area to focus on\n",
    "fig, ax = plt.subplots()\n",
    "ax.grid(False)\n",
    "ax.scatter(df['x_centroid'], df['y_centroid'], s=0.1, alpha=0.1, color='lightgray')\n",
    "mask = (df['x_centroid'] < 7000) & (df['x_centroid'] > 6500)\n",
    "mask = mask & (df['y_centroid'] < 8500) & (df['y_centroid'] > 8000)\n",
    "ax.scatter(df.loc[mask, 'x_centroid'], df.loc[mask, 'y_centroid'], s=0.1, alpha=0.1, color='r')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0fa5ca99-43f5-4c74-89da-da791e4f596c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# add spatial and subset\n",
    "avis.obsm['spatial'] = df.loc[avis.obs.index, ['x_centroid','y_centroid']].values\n",
    "avis = avis[mask].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a75eff20-8f37-49be-bdab-ba18c1c4ada6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# compute PCA for the spatial dataset using HVGs\n",
    "avis.uns['log1p']['base'] = np.e\n",
    "sc.pp.highly_variable_genes(avis, flavor='seurat', min_mean=0.5, max_mean=7.5, min_disp=0.5, n_top_genes=avis.shape[1])\n",
    "sc.pl.highly_variable_genes(avis)\n",
    "# report how many HVGs we're going to use\n",
    "print('there are', avis.var['highly_variable'].sum(), 'hvgs')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "91b57da8-514e-482a-a178-9bfc93baaf70",
   "metadata": {},
   "source": [
    "## learn with CCE for xenium"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4828de8e-68f8-453a-b261-199d4b416579",
   "metadata": {},
   "outputs": [],
   "source": [
    "from scipy.spatial.distance import pdist, squareform\n",
    "# create a gene subset for testing\n",
    "genes = avis.var_names[avis.var['highly_variable']].tolist()\n",
    "# retrieve positions from the data\n",
    "pos = torch.from_numpy(df[['x_centroid','y_centroid']].values)\n",
    "# convert the gene expression data to numpy\n",
    "x = torch.from_numpy(avis[:, genes].X.todense())\n",
    "# sphere normalize the data (just in case)\n",
    "norm_factor = torch.pow(x, 2).sum(1).reshape(-1,1)\n",
    "assert (norm_factor > 0).all()\n",
    "x = torch.div(x, norm_factor)\n",
    "# compute the distances\n",
    "distances = squareform(pdist(df.loc[mask, ['x_centroid','y_centroid']]))\n",
    "# compute the edges as two cell widths apart so 30µm\n",
    "edge_index = torch.from_numpy(np.array(np.where((distances < 15)&(distances != 0))))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9dbf957f-3c07-4654-bed8-fd6258d34295",
   "metadata": {},
   "outputs": [],
   "source": [
    "# define the parameters of the model\n",
    "n_genes = len(genes)\n",
    "learning_rate = 5e-2\n",
    "zmft_scalar = 1e-1\n",
    "seed = 0\n",
    "epochs = 200"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1c640680-21dd-4a2c-b2ea-0a1489eaea16",
   "metadata": {},
   "outputs": [],
   "source": [
    "# instantiate the model, input and output will be the same\n",
    "model_rev = celcomen(input_dim=n_genes, output_dim=n_genes, n_neighbors=n_neighbors, seed=seed)\n",
    "\n",
    "# now perform the simulation\n",
    "np.random.seed(seed)\n",
    "# artifically set the g2g matrix\n",
    "input_g2g = np.random.uniform(size=(n_genes, n_genes)).astype('float32')\n",
    "input_g2g = normalize_g2g((input_g2g + input_g2g.T) / 2)\n",
    "model_rev.set_g2g(torch.from_numpy(input_g2g))\n",
    "model_rev.set_g2g_intra(torch.from_numpy(input_g2g))\n",
    "# initialize a gene expression matrix\n",
    "model_rev.set_gex(x)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f5562e81-bd73-496e-bf6d-5ff89082f47b",
   "metadata": {},
   "outputs": [],
   "source": [
    "from tqdm import tqdm\n",
    "# setup the initial optimizer\n",
    "optimizer = torch.optim.SGD(model_rev.parameters(), lr=learning_rate, momentum=0)\n",
    "# keep track of the losses per data object\n",
    "loss, losses = None, []\n",
    "# train the model\n",
    "model_rev.train()\n",
    "# work through epochs\n",
    "for epoch in tqdm(range(epochs), total=epochs):\n",
    "    # derive the message as well as the mean field approximation\n",
    "    msg, msg_intra, log_z_mft = model_rev(edge_index, 1)\n",
    "    # compute the loss and track it\n",
    "    loss = -(-log_z_mft + zmft_scalar * torch.trace(torch.mm(msg, torch.t(model_rev.gex))) + zmft_scalar * torch.trace(torch.mm(msg_intra, torch.t(model_rev.gex))) )\n",
    "    losses.append(loss.detach().numpy()[0][0])\n",
    "    # derive the gradients, update, and clear\n",
    "    loss.backward()\n",
    "    optimizer.step()\n",
    "    optimizer.zero_grad()\n",
    "    # repeatedly force a normalization\n",
    "    model_rev.conv1.lin.weight = torch.nn.Parameter(normalize_g2g(model_rev.conv1.lin.weight), requires_grad=True)\n",
    "    model_rev.lin.weight = torch.nn.Parameter(normalize_g2g(model_rev.lin.weight), requires_grad=True)\n",
    "    optimizer = torch.optim.SGD(model_rev.parameters(), lr=learning_rate, momentum=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1447cb11-1a5a-4e4d-b79e-8f645af7b804",
   "metadata": {},
   "outputs": [],
   "source": [
    "# create the plot\n",
    "fig, ax = plt.subplots(figsize=[6, 4])\n",
    "ax.grid(False)\n",
    "ax.plot(losses, lw=2, color='#fe86a4')\n",
    "ax.set_xlim(0, epochs)\n",
    "vmin, vmax = min(min(losses), 0), max(losses)\n",
    "vstep = (vmax - vmin) * 0.01\n",
    "ax.set_ylim(vmin-vstep, vmax+vstep)\n",
    "ax.set(xlabel='epochs', ylabel='loss')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3bce5495-aab6-4a5f-8581-3be0ca3e5c2c",
   "metadata": {},
   "source": [
    "## perform disentanglement"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "59ae8034-7f06-4ae3-8050-7ccaa0b31b65",
   "metadata": {},
   "outputs": [],
   "source": [
    "# retrieve the g2gs\n",
    "g2g_intercellular = pd.DataFrame(model_rev.conv1.lin.weight.detach().numpy(), index=genes, columns=genes)\n",
    "g2g_intracellular = pd.DataFrame(model_rev.lin.weight.detach().numpy(), index=genes, columns=genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5c4b47f8-b778-4edc-9ae8-d0d733ad6c9f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# define the gene\n",
    "secs = pd.read_table('quickgo_secreted.tsv')['SYMBOL'].unique()\n",
    "sps = pd.read_table('quickgo_surfaceprotein.tsv')['SYMBOL'].unique()\n",
    "nons = pd.Index(genes)[(~pd.Index(genes).isin(secs))&(~pd.Index(genes).isin(sps))]\n",
    "secs = pd.Index(genes).intersection(secs)\n",
    "sps = pd.Index(genes).intersection(sps)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2b9cc1ed-9f21-4768-b87a-3a88cf5d1bc2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# derive the values\n",
    "diffs_secs = (g2g_intercellular.loc[secs, :] - g2g_intracellular.loc[secs, :]).values.flatten()\n",
    "diffs_sps = (g2g_intercellular.loc[sps, :] - g2g_intracellular.loc[sps, :]).values.flatten()\n",
    "diffs_nons = (g2g_intercellular.loc[nons, :] - g2g_intracellular.loc[nons, :]).values.flatten()\n",
    "# assemble them\n",
    "ys = np.hstack([diffs_secs, diffs_sps, diffs_nons])\n",
    "ys -= ys.mean()\n",
    "ys /= ys.std()\n",
    "xs = np.array(['Secreted']*len(diffs_secs)+['Membrane']*len(diffs_sps)+['Cytoplasmic']*len(diffs_nons))\n",
    "# create plot\n",
    "fig, ax = plt.subplots(figsize=[2, 4])\n",
    "ax.grid(False)\n",
    "sns.barplot(x=xs[::-1], y=ys[::-1], saturation=1, ci=95, errwidth=2, capsize=0.3, linewidth=2,\n",
    "            errcolor='k', edgecolor='k', palette=['#cc78bc','#d55e00','#029e73'][::-1])\n",
    "ax.set_xlim(-0.75, 2.75)\n",
    "ax.tick_params(axis='x', labelrotation=90)\n",
    "ax.set_ylabel('Inter - intra G2G (Z-score)')\n",
    "ax.spines[['right', 'top']].set_visible(False)\n",
    "# derive pvalues\n",
    "p1 = ss.mannwhitneyu(ys[xs == 'Secreted'], ys[xs == 'Membrane'])[1]\n",
    "p2 = ss.mannwhitneyu(ys[xs == 'Secreted'], ys[xs == 'Cytoplasmic'])[1]\n",
    "p3 = ss.mannwhitneyu(ys[xs == 'Cytoplasmic'], ys[xs == 'Membrane'])[1]\n",
    "print(p1, p2, p3, sep='\\n')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9ad41e9f-2a01-40b9-a8c3-07ccd139940a",
   "metadata": {},
   "source": [
    "## perturb the GBM data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d53557eb-2a0f-4133-953e-3467e682a623",
   "metadata": {},
   "outputs": [],
   "source": [
    "# define a function to derive the gex from the sphex\n",
    "def calc_gex(sphex):\n",
    "    \"\"\"\n",
    "    Calculates the gene expression matrix from the spherical\n",
    "    \"\"\"\n",
    "    # setup the gex\n",
    "    n_genes = sphex.shape[1]+1\n",
    "    gex = torch.from_numpy(np.zeros((sphex.shape[0], n_genes)).astype('float32'))\n",
    "    # compute the gex\n",
    "    for idx in range(n_genes):\n",
    "        if idx == n_genes-1:\n",
    "            gex[:,idx] = torch.sin(sphex[:,idx-1])\n",
    "        else:\n",
    "            gex[:,idx] = torch.cos(sphex[:,idx])\n",
    "        for idx_ in range(idx):\n",
    "            gex[:,idx] *= torch.sin(sphex[:,idx_])\n",
    "    return torch.nan_to_num(gex)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5fa78267-905b-42d2-afc8-e7e5697b1cf3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# define a function to derive the gex from the sphex\n",
    "def calc_sphex(gex):\n",
    "    \"\"\"\n",
    "    Calculates the spherical expression matrix from the normal\n",
    "    \"\"\"\n",
    "    # setup the gex\n",
    "    n_sgenes = gex.shape[1]-1\n",
    "    sphex = torch.from_numpy(np.zeros((gex.shape[0], n_sgenes)).astype('float32'))\n",
    "    # compute the gex\n",
    "    for idx in range(n_sgenes):\n",
    "        sphex[:,idx] = gex[:,idx]\n",
    "        for idx_ in range(idx):\n",
    "            sphex[:,idx] /= torch.sin(sphex[:,idx_])\n",
    "        sphex[:,idx] = torch.arccos(sphex[:,idx])\n",
    "    return sphex"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7fa012a7-1f73-4e50-b657-a275d4f078db",
   "metadata": {},
   "outputs": [],
   "source": [
    "# define the celcomen class\n",
    "class simcomen(torch.nn.Module):\n",
    "    # define initialization function\n",
    "    def __init__(self, input_dim, output_dim, n_neighbors, seed=0):\n",
    "        super(simcomen, self).__init__()\n",
    "        # define the seed\n",
    "        torch.manual_seed(seed)\n",
    "        # set up the graph convolution\n",
    "        self.conv1 = GCNConv(input_dim, output_dim, add_self_loops=False)\n",
    "        # set up the linear layer for intracellular gene regulation\n",
    "        self.lin = torch.nn.Linear(input_dim, output_dim)\n",
    "        # define the neighbors\n",
    "        self.n_neighbors = n_neighbors\n",
    "        # define a tracking variable for the gene expression x matrix\n",
    "        self.sphex = None\n",
    "        self.gex = None\n",
    "\n",
    "    # define a function to artificially set the g2g matrix\n",
    "    def set_g2g(self, g2g):\n",
    "        \"\"\"\n",
    "        Artifically sets the core g2g matrix to be a specified interaction matrix\n",
    "        \"\"\"\n",
    "        # set the weight as the input\n",
    "        self.conv1.lin.weight = torch.nn.Parameter(g2g, requires_grad=False)\n",
    "        # and then set the bias as all zeros\n",
    "        self.conv1.bias = torch.nn.Parameter(torch.from_numpy(np.zeros(n_genes).astype('float32')), requires_grad=False)\n",
    "\n",
    "    # define a function to artificially set the g2g matrix\n",
    "    def set_g2g_intra(self, g2g_intra):\n",
    "        \"\"\"\n",
    "        Artifically sets the core g2g intracellular matrix to be a specified matrix\n",
    "        \"\"\"\n",
    "        # set the weight as the input\n",
    "        self.lin.weight = torch.nn.Parameter(g2g_intra, requires_grad=False)\n",
    "        # and then set the bias as all zeros\n",
    "        self.lin.bias = torch.nn.Parameter(torch.from_numpy(np.zeros(len(g2g_intra)).astype('float32')), requires_grad=False)\n",
    "\n",
    "    # define a function to artificially set the sphex matrix\n",
    "    def set_sphex(self, sphex):\n",
    "        \"\"\"\n",
    "        Artifically sets the current sphex matrix\n",
    "        \"\"\"\n",
    "        self.sphex = torch.nn.Parameter(sphex, requires_grad=True)\n",
    "        \n",
    "    # define the forward pass\n",
    "    def forward(self, edge_index, batch):\n",
    "        \"\"\"\n",
    "        Forward pass for prediction or training,\n",
    "        convolutes the input by the expected interactions and returns log(Z_mft)\n",
    "        \"\"\"\n",
    "        # compute the gex\n",
    "        self.gex = calc_gex(self.sphex)\n",
    "        # compute the message\n",
    "        msg = self.conv1(self.gex, edge_index)\n",
    "        # compute intracellular message\n",
    "        msg_intra = self.lin(self.gex)\n",
    "        # compute the log z mft\n",
    "        log_z_mft = self.log_Z_mft(edge_index, batch)\n",
    "        return msg, msg_intra, log_z_mft\n",
    "\n",
    "    # define approximation function\n",
    "    def log_Z_mft(self, edge_index, batch):\n",
    "        \"\"\"\n",
    "        Mean Field Theory approximation to the partition function. Assumptions used are:\n",
    "        - expression of values of genes are close to their mean values over the visium slide\n",
    "        - \\sum_b g_{a,b} m^b >0 \\forall a, where m is the mean gene expression and g is the gene-gene\n",
    "          interaction matrix.\n",
    "        \"\"\"\n",
    "        # retrieve number of spots\n",
    "        num_spots = self.gex.shape[0]\n",
    "        # calculate mean gene expression        \n",
    "        mean_genes = torch.mean(self.gex, axis=0).reshape(-1,1)  # the mean should be per connected graph\n",
    "        # calculate the norm of the sum of mean genes\n",
    "        g = torch.norm(torch.mm( self.n_neighbors*self.conv1.lin.weight + 2*self.lin.weight, mean_genes))   # maybe needs to change to g = torch.norm(torch.mm(mean_genes, self.conv1.lin.weight))\n",
    "        # calculate the contribution for mean values        \n",
    "        z_mean = - num_spots  * torch.mm(torch.mm(torch.t(mean_genes), self.lin.weight + 0.5 * self.n_neighbors * self.conv1.lin.weight),  mean_genes)\n",
    "        # calculate the contribution gene interactions\n",
    "        z_interaction = self.z_interaction(num_spots=num_spots, g=g)\n",
    "        # add the two contributions        \n",
    "        log_z_mft = z_mean + z_interaction\n",
    "        return log_z_mft\n",
    "\n",
    "    def z_interaction(self, num_spots, g):\n",
    "        \"\"\"\n",
    "        Avoid exploding exponentials by returning an approximate interaction term for the partition function.\n",
    "        \"\"\"\n",
    "        if g>20:\n",
    "            z_interaction = num_spots * ( g - torch.log( g) )\n",
    "        else:\n",
    "            z_interaction = num_spots * torch.log((torch.exp( g) - torch.exp(- g))/( g))\n",
    "        return z_interaction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4619a957-ee1d-4b5b-bdd4-6153de5432ed",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# find the genes\n",
    "avis_sub = avis[:, avis.var['highly_variable']].copy()\n",
    "sc.pl.embedding(avis_sub, basis='spatial', color=['IFIT2','IFIT3','IFITM3','IFNB1'], use_raw=False, s=1e2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "75e3db27-0573-4806-8d10-94f105d7d1c8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# define the genes that distinguish ISG15 spots currently\n",
    "avis_sub.obs['groups'] = (sc.get.obs_df(avis_sub, keys=['IFITM3']).max(1) > 0).map({True:'IFITM3+', False:'-'})\n",
    "sc.pl.embedding(avis_sub, basis='spatial', color=['groups'], use_raw=False, s=1e2)\n",
    "# calculate differential expression\n",
    "sc.tl.rank_genes_groups(avis_sub, method='wilcoxon', use_raw=False, groupby='groups', group=['IFITM3+'], reference='-')\n",
    "sc.pl.rank_genes_groups(avis_sub, n_genes=25, sharey=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f0565e5a-b818-4b71-b356-f72c3caf22de",
   "metadata": {},
   "outputs": [],
   "source": [
    "# derive the signature\n",
    "'\\''+'\\',\\''.join(pd.DataFrame(avis_sub.uns['rank_genes_groups']['names']).iloc[:25, 0].tolist())+'\\''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e850131f-76e6-47ee-8031-6e1052fb5982",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# propose an X introducing ISG15 inflammation\n",
    "proposed_x = avis_sub.X.toarray().copy()\n",
    "# adjust the X so we artificially introduce signaling to the center left side of the tissue\n",
    "np.random.seed(0)\n",
    "df_gex = sc.get.obs_df(avis_sub, keys=['IFITM3'])['IFITM3']\n",
    "mask = avis_sub.obs.index == np.random.choice(avis_sub.obs.index[df_gex > 0], size=1)[0]\n",
    "idx = np.where(avis_sub.var_names == 'IFITM3')[0][0]\n",
    "proposed_x[mask, idx] = 0\n",
    "avis_sub.obs['perturbed'] = 'unperturbed'\n",
    "avis_sub.obs['perturbed'].iloc[np.where(squareform(pdist(avis_sub.obsm['spatial']))[mask, :] < 30)[1]] = 'perturbed-neighbors1'\n",
    "avis_sub.obs['perturbed'].iloc[np.where(squareform(pdist(avis_sub.obsm['spatial']))[mask, :] < 15)[1]] = 'perturbed-neighbors0'\n",
    "avis_sub.obs.loc[mask, 'perturbed'] = 'perturbed'\n",
    "avis_sub.uns['perturbed_colors'] = ['#ff47a6','#f593c2','#f7cbe0','#f7ebf1']\n",
    "sc.pl.embedding(avis_sub, basis='spatial', color=['perturbed'], use_raw=False, s=1e2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0042c824-79ca-4703-8e12-a015870b9acd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# demonstrate what the proposed x looks like pre scaling\n",
    "avis_sub.X = proposed_x\n",
    "sc.pl.embedding(avis_sub, basis='spatial', color=['IFITM3','IFIT3','MX1','IRF1'], use_raw=False, s=1e2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "78bc7ccc-a210-4f09-98bd-6cd4825fd1bc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# define the parameters of the model\n",
    "n_genes = avis_sub.shape[1]\n",
    "learning_rate = 1e-3\n",
    "zmft_scalar = 1e-1\n",
    "seed = 0\n",
    "epochs = 50"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "456c07cb-89e9-46f3-8a85-196f9fff97b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# instantiate the model, input and output will be the same\n",
    "model = simcomen(input_dim=n_genes, output_dim=n_genes, n_neighbors=n_neighbors, seed=seed)\n",
    "# now perform the simulation\n",
    "np.random.seed(seed)\n",
    "# convert the gene expression data to numpy\n",
    "x = torch.from_numpy(proposed_x)\n",
    "# sphere normalize the data (just in case)\n",
    "norm_factor = torch.sqrt(torch.pow(x, 2).sum(1)).reshape(-1,1)\n",
    "assert (norm_factor > 0).all()\n",
    "x = torch.div(x, norm_factor)\n",
    "# artifically set the g2g matrix\n",
    "model.set_g2g(model_rev.conv1.lin.weight.clone().detach())\n",
    "model.set_g2g_intra(model_rev.lin.weight.clone().detach())\n",
    "# initialize a gene expression matrix\n",
    "input_sphex = calc_sphex(x).clone().detach().numpy()\n",
    "model.set_sphex(torch.from_numpy(input_sphex.astype('float32')))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1a4368f7-bb0a-42ed-bcd0-63b2a3bf1a3e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# set up the optimizer\n",
    "optimizer = torch.optim.SGD(model.parameters(), lr=learning_rate, momentum=0)\n",
    "# keep track of the losses per data object\n",
    "loss, losses = None, []\n",
    "# train the model\n",
    "model.train()\n",
    "tmp_gexs = []\n",
    "# work through epochs\n",
    "for epoch in tqdm(range(epochs), total=epochs):\n",
    "    # derive the message as well as the mean field approximation\n",
    "    msg, msg_intra, log_z_mft = model(edge_index, 1)\n",
    "    if (epoch % 5) == 0:\n",
    "        tmp_gex = model.gex.clone().detach().numpy()\n",
    "        tmp_gexs.append(tmp_gex)\n",
    "    # compute the loss and track it\n",
    "    loss = -(-log_z_mft + zmft_scalar * torch.trace(torch.mm(msg, torch.t(model.gex))) + zmft_scalar * torch.trace(torch.mm(msg_intra, torch.t(model.gex))) )\n",
    "    losses.append(loss.detach().numpy()[0][0])\n",
    "    # derive the gradients, update, and clear\n",
    "    loss.backward()\n",
    "    optimizer.step()\n",
    "    optimizer.zero_grad()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "61b0c009-4ae4-4abf-8186-ee46a7140c7e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot the loss\n",
    "fig, ax = plt.subplots(figsize=[6, 4])\n",
    "ax.grid(False)\n",
    "ax.plot(losses, lw=2, color='#fe86a4')\n",
    "ax.set_xlim(0, epochs)\n",
    "vmin, vmax = min(min(losses), 0), max(losses)\n",
    "vstep = (vmax - vmin) * 0.01\n",
    "ax.set_ylim(vmin-vstep, vmax+vstep)\n",
    "ax.set(xlabel='epochs', ylabel='loss')\n",
    "# retrieve the data\n",
    "output_gex = model.gex.detach().numpy()\n",
    "output_msg = msg.detach().numpy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2ae57274-b6dd-45f4-aa4a-28389a21251f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# find the genes of interest\n",
    "genes = ['IFITM3','IFIT3','MX1','IRF1']\n",
    "genes = [x for x in genes if x in avis_sub.var_names.tolist()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "778524b4-bff5-4dde-bc90-fa7f7f44b7fc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# add the layers\n",
    "for idx, tmp_gex in enumerate(tmp_gexs):\n",
    "    avis_sub.layers[f'input{idx}'] = tmp_gex\n",
    "avis_sub.layers['output'] = output_gex"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0bb4937b-04d9-43a7-8061-816dbb919ba3",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# plot the genes of interest\n",
    "sc.pl.embedding(avis_sub, basis='spatial', color=genes, use_raw=False, s=1e2, layer='input0')\n",
    "sc.pl.embedding(avis_sub, basis='spatial', color=genes, use_raw=False, s=1e2, layer='input3')\n",
    "sc.pl.embedding(avis_sub, basis='spatial', color=genes, use_raw=False, s=1e2, layer='output')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4096fbcc-bf1a-422d-8a11-c7c4be43efc0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# compute the differential within the area of interest vs outside\n",
    "avis_sub.layers['diff'] = avis_sub.layers['output'] - avis_sub.layers['input0']\n",
    "mask0 = avis_sub.obs['perturbed'] == 'unperturbed'\n",
    "mask1 = avis_sub.obs['perturbed'] == 'perturbed'\n",
    "diff_in_vs_out_p = pd.Series((avis_sub.layers['diff'][mask1].mean(0) - avis_sub.layers['diff'][mask0].mean(0)),\n",
    "                             index=avis_sub.var_names).sort_values()[::-1]\n",
    "# now compute differential for neighbors\n",
    "mask1 = avis_sub.obs['perturbed'] == 'perturbed-neighbors0'\n",
    "diff_in_vs_out_pn0 = pd.Series((avis_sub.layers['diff'][mask1].mean(0) - avis_sub.layers['diff'][mask0].mean(0)),\n",
    "                             index=avis_sub.var_names).sort_values()[::-1]\n",
    "mask1 = avis_sub.obs['perturbed'] == 'perturbed-neighbors1'\n",
    "diff_in_vs_out_pn1 = pd.Series((avis_sub.layers['diff'][mask1].mean(0) - avis_sub.layers['diff'][mask0].mean(0)),\n",
    "                             index=avis_sub.var_names).sort_values()[::-1]\n",
    "# remove the perturbed genes to get a clean look\n",
    "perturbed_genes = ['IFITM3']\n",
    "perturbed_label = 'ifitm3'\n",
    "diff_in_vs_out_p = diff_in_vs_out_p.loc[~diff_in_vs_out_p.index.isin(perturbed_genes)]\n",
    "diff_in_vs_out_pn0 = diff_in_vs_out_pn0.loc[~diff_in_vs_out_pn0.index.isin(perturbed_genes)]\n",
    "diff_in_vs_out_pn1 = diff_in_vs_out_pn1.loc[~diff_in_vs_out_pn1.index.isin(perturbed_genes)]\n",
    "# write these matrices out\n",
    "diff_in_vs_out_p.to_csv(f'20240618.learn_then_perturb.xenium_gbm.{perturbed_label}.diff_in_vs_out_p.csv')\n",
    "diff_in_vs_out_pn0.to_csv(f'20240618.learn_then_perturb.xenium_gbm.{perturbed_label}.diff_in_vs_out_pn0.csv')\n",
    "diff_in_vs_out_pn1.to_csv(f'20240618.learn_then_perturb.xenium_gbm.{perturbed_label}.diff_in_vs_out_pn1.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "827a43ba-415f-4db1-afb1-17cd18c57a96",
   "metadata": {},
   "outputs": [],
   "source": [
    "# do not include isg15 itself\n",
    "genes_ = pd.DataFrame(avis_sub.uns['rank_genes_groups']['names']).iloc[:25, 0].tolist()[1:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6fdfb06c-4fd0-4656-aea4-a5018b8e3ae5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# retrieve the results\n",
    "df_expr = pd.DataFrame(columns=['p','pn0','pn1','u'])\n",
    "for layer in ['input1','input2','input3','input4','input5','input6','input7','input8','input9','output']:\n",
    "    # compute the expression\n",
    "    mask1 = avis_sub.obs['perturbed'] == 'perturbed'\n",
    "    expr_p = pd.Series(avis_sub.layers[layer][mask1].mean(0), index=avis_sub.var_names).loc[genes_].mean()\n",
    "    mask1 = avis_sub.obs['perturbed'] == 'perturbed-neighbors0'\n",
    "    expr_pn0 = pd.Series(avis_sub.layers[layer][mask1].mean(0), index=avis_sub.var_names).loc[genes_].mean()\n",
    "    mask1 = avis_sub.obs['perturbed'] == 'perturbed-neighbors1'\n",
    "    expr_pn1 = pd.Series(avis_sub.layers[layer][mask1].mean(0), index=avis_sub.var_names).loc[genes_].mean()\n",
    "    mask1 = avis_sub.obs['perturbed'] == 'unperturbed'\n",
    "    expr_u = pd.Series(avis_sub.layers[layer][mask1].mean(0), index=avis_sub.var_names).loc[genes_].mean()\n",
    "    df_expr.loc[layer] = expr_p, expr_pn0, expr_pn1, expr_u"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a5cd183d-c434-485f-96de-899605072014",
   "metadata": {},
   "outputs": [],
   "source": [
    "# derive the scores\n",
    "fig, ax = plt.subplots(figsize=[2.9, 5.6])\n",
    "ax = sns.heatmap(df_expr, cmap='viridis')\n",
    "ax.grid(False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3f8c55a0-19e5-4841-8f32-5cdd13c26d13",
   "metadata": {},
   "outputs": [],
   "source": [
    "# derive the scores\n",
    "df_expr -= np.mean(df_expr.values.flatten())\n",
    "df_expr /= np.std(df_expr.values.flatten())\n",
    "fig, ax = plt.subplots(figsize=[2.9, 5.6])\n",
    "ax = sns.heatmap(df_expr, cmap='viridis')\n",
    "ax.grid(False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8aebca86-f7ab-4135-aae6-b79329f11120",
   "metadata": {},
   "outputs": [],
   "source": [
    "# derive the scores\n",
    "for col, color in zip(['p','pn0','pn1','u'], ['#ff47a6','#f593c2','#f7cbe0','#f7ebf1']):\n",
    "    fig, ax = plt.subplots(figsize=[4, 4])\n",
    "    ax.scatter(df_expr.index, df_expr[col], color='k', s=9e1, zorder=1)\n",
    "    ax.scatter(df_expr.index, df_expr[col], color=color, s=5e1, zorder=2)\n",
    "    ax.plot(df_expr.index, df_expr[col], color='k', lw=3.5, zorder=0)\n",
    "    ax.plot(df_expr.index, df_expr[col], color=color, lw=1, zorder=2)\n",
    "    ax.grid(False)\n",
    "    ax.tick_params(axis='x', labelrotation=90)\n",
    "    ax.set(xlabel='timepoint', ylabel='IFITM3 score', title=col)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4aa68448-c247-4db2-acac-e3d7b0cf268c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # run the following in R\n",
    "# # load gene sets\n",
    "# # > import package\n",
    "# library(msigdbr)\n",
    "# print(paste0('MSigDb Version: ', packageVersion(\"msigdbr\")))\n",
    "# # > gather gene sets\n",
    "# categories <- c(\"H\", \"C2\", \"C5\", \"C6\", \"C7\")\n",
    "# gene_sets <- c()  # instantiate list of all gene sets\n",
    "# msigdbr_t2gs <- c()\n",
    "# for (category in categories) {\n",
    "#   msigdbr_df <- msigdbr(species = \"Homo sapiens\", category = category)\n",
    "#   msigdbr_list <- split(x = msigdbr_df$gene_symbol, f = msigdbr_df$gs_name)\n",
    "#   gene_sets <- c(gene_sets, msigdbr_list)\n",
    "#   msigdbr_t2g <- msigdbr_df %>% dplyr::distinct(gs_name, gene_symbol) %>% as.data.frame()\n",
    "#   msigdbr_t2gs <- rbind(msigdbr_t2gs, msigdbr_t2g)\n",
    "# }\n",
    "# \n",
    "# # retrieve the gsea object\n",
    "# library(GSEABase)\n",
    "# library(clusterProfiler)\n",
    "# library(DOSE)\n",
    "# library(org.Mm.eg.db)\n",
    "# library(ggplot2)\n",
    "# library(stringr)\n",
    "# \n",
    "# # add on our custom signature\n",
    "# genes <- c('IGFBP7','C1R','C1S','A2M','APOE','CD44','IFIT3','SOCS3','FN1','LAMP2','SPARCL1','PTGDS','DST','ANXA1','MX1','STAT1','CD38','STAT2','EDNRB','NOTCH3','IRF1','CSF1','HLA-B','STAT3')\n",
    "# tmp <- data.frame(gs_name='DC20_CUSTOM_IFITM3_SIGNATURE', gene_symbol=genes)\n",
    "# msigdbr_t2gs_tmp <- rbind(msigdbr_t2gs, tmp)\n",
    "# \n",
    "# # define the filename\n",
    "# for (comp in c(\"diff_in_vs_out_p\", \"diff_in_vs_out_pn0\", \"diff_in_vs_out_pn1\")) {\n",
    "#   fn <- paste0(\"20240618.learn_then_perturb.xenium_gbm.ifitm3.\", comp, \".csv\")\n",
    "#   # read in the rankings\n",
    "#   rankedGenes <- read.csv(fn, row.names=1)\n",
    "#   # rename and shape the rankings\n",
    "#   geneNames <- rownames(rankedGenes)\n",
    "#   geneList <- rankedGenes$X0\n",
    "#   names(geneList) <- geneNames\n",
    "#   geneList <- sort(geneList, decreasing=T)\n",
    "#   # create signatures\n",
    "#   egmt <- GSEA(geneList, TERM2GENE=msigdbr_t2gs, verbose=F, pvalueCutoff=1,\n",
    "#                minGSSize=15, maxGSSize=500)\n",
    "#   # save the data\n",
    "#   data <- data.frame(egmt)[c(\"enrichmentScore\",\"NES\",\"pvalue\",\"p.adjust\",\"qvalues\",\"setSize\")]\n",
    "#   out_fn <- paste0(\"20240618.learn_then_perturb.xenium_gbm.ifitm3.\", comp, \".gsea.csv\")\n",
    "#   write.csv(data, file=out_fn)\n",
    "# }"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Environment (base_py39)",
   "language": "python",
   "name": "base_py39"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
